### PR TITLE
8304069: ClassFileParser has ad-hoc hashtables

### PR DESCRIPTION
**Pull request by matias9927**
The ClassFileParser has ad-hoc hash tables for checking for duplicate names for fields, methods and interfaces. ResourceHashtable will now be used instead. Verified with tier1-4 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304069](https://bugs.openjdk.org/browse/JDK-8304069): ClassFileParser has ad-hoc hashtables

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13141/head:pull/13141` \
`$ git checkout pull/13141`

Update a local copy of the PR: \
`$ git checkout pull/13141` \
`$ git pull https://git.openjdk.org/jdk.git pull/13141/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13141`

View PR using the GUI difftool: \
`$ git pr show -t 13141`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13141.diff">https://git.openjdk.org/jdk/pull/13141.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -882,27 +857,23 @@ void ClassFileParser::parse_interfaces(const ClassFileStream* const stream,
...
+    NameSigHashtable* interface_names = new NameSigHashtable();
+    bool dup = true;
+    NameSigHash* interface_name = nullptr;
     {
       debug_only(NoSafepointVerifier nsv;)
-      for (index = 0; index < itfs_len; index++) {
+      for (index = 0; (index < itfs_len) && dup; index++) {
         const InstanceKlass* const k = _local_interfaces->at(index);
-        name = k->name();
+        interface_name = new NameSigHash();
```
Comment (coleenp) [inline src/hotspot/share/classfile/classFileParser.cpp:868]: You could use a local NameSigHash and copy for the values like:
  name_sig_hash = NameSigHash(k->name(), nullptr);
Then the hashtable key could be a non-pointer, saving a pointer per entry and may be more efficient.

**Review by coleenp — CHANGES_REQUESTED** [review]
I have some suggestions for improving this.

**Review by coleenp — APPROVED** [review]
This looks great.

**Review by dholmes-ora — COMMENTED** [review]
This conversion looks good but I was confused by the way the original code was doing things. Once we have this conversion I think we could clean things up further:

1. Interfaces don't need `NameSigHash`, all we need is a map of names, so we should be able to directly have a RHT of `Symbol*` for the `interface_name_table`.
2. `NameSigHash` is the wrong terminology - we are checking names and descriptors [JVMS 4.3], not names and signatures [JVMS4.7.9.1]! (And a Java Language "signature" is something different again [JLS 8.4.2]!

Unfortunately I think we misuse "signature" throughout the VM, so maybe fixing it here will serve little purpose.

**Review by coleenp — CHANGES_REQUESTED** [review]
Every cleanup suggests more cleanups.

**Review by coleenp — APPROVED** [review]
Looks good to me!

**Review by dholmes-ora — APPROVED** [review]
Looks good! Thanks for the additional cleanups/simplifications.

**Pull request closed without merging**